File: Data_Structures_And_Algorithms-UC_SanDiego/1-Algorithmic_Toolbox/Week2/Programming_Assignment/5_fibonacci_number_again/fibonacci_huge.py
# Uses python3
import sys

def pisanoPeriod(m):
    previous, current = 0, 1
    for i in range(0, m * m):
        previous, current \
        = current, (previous + current) % m
         
        # A Pisano Period starts with 01
        if (previous == 0 and current == 1):
            return i + 1
 
# Calculate Fn mod m 
def get_fibonacci_huge(n, m):
     
    # Getting the period
    pisano_period = pisanoPeriod(m)
    # Taking mod of N with 
    # period length
    n = n % pisano_period
     
    previous, current = 0, 1
    if n==0:
        return 0
    elif n==1:
        return 1
    for i in range(n-1):
        previous, current \
        = current, previous + current
         
    return (current % m)

if __name__ == '__main__':
    n, m = map(int, input().split())
    print(get_fibonacci_huge(n, m))

File: Data_Structures_And_Algorithms-UC_SanDiego/1-Algorithmic_Toolbox/Week2/Programming_Assignment/5_fibonacci_number_again/fibonacci_huge.cpp
#include <iostream>

int pisanoPeriod(int m){
    int previous = 0;
    int current = 1;
    for(int i=0; i!=m*m; i++){
        previous = current;
        current = (int)(previous+current)%m;
        if (previous==0 and current==1){
            return i+1;
        }
    }
}

long long get_fibonacci_huge(long long n, long long m) {
    if (n <= 1)
        return n;
    n = (long long)n%pisanoPeriod(m);
    long long previous = 0;
    long long current  = 1;

    for (long long i = 0; i < (n - 1); ++i) {
        std::cout << current << std::endl;
        previous = current;
        current = previous + current;
    }

    return current % m;
}

int main() {
    long long n, m;
    std::cin >> n >> m;
    std::cout << get_fibonacci_huge(n, m) << '\n';
}
